## AIND - Planning Lab Analysis 


This notebook intends to make an analysis in order to fully complete AIND Planning Lab.

### Part 2 - Domain-independent heuristics:

After coding `Planning Graph`, it'll be presented results running different search methods.




### First problem: air_cargo_p1

- Search methods: astar_search h_1, astar_search h_ignore_preconditions and astar_search h_pg_levelsum

### 1) - astar_search h_1

In [2]:
%run run_search.py -p 1 -s 8


Solving Air Cargo Problem 1 using astar_search with h_1...

Expansions   Goal Tests   New Nodes
    55          57         224    

Plan length: 6  Time elapsed in seconds: 0.0789819739293307
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)



### 2) - astar_search h_ignore_preconditions

In [3]:
%run run_search.py -p 1 -s 9


Solving Air Cargo Problem 1 using astar_search with h_ignore_preconditions...

Expansions   Goal Tests   New Nodes
    41          43         170    

Plan length: 6  Time elapsed in seconds: 0.08492878801189363
Load(C1, P1, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)



### 3) - astar_search h_pg_levelsum

In [4]:
%run run_search.py -p 1 -s 10


Solving Air Cargo Problem 1 using astar_search with h_pg_levelsum...

Expansions   Goal Tests   New Nodes
    11          13          50    

Plan length: 6  Time elapsed in seconds: 3.08339868998155
Load(C1, P1, SFO)
Fly(P1, SFO, JFK)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)



### Second problem: air_cargo_p2

- Search methods: astar_search h_1, astar_search h_ignore_preconditions and astar_search h_pg_levelsum

### 1) - astar_search h_1

In [6]:
%run run_search.py -p 2 -s 8


Solving Air Cargo Problem 2 using astar_search with h_1...

Expansions   Goal Tests   New Nodes
   4853        4855       44041   

Plan length: 9  Time elapsed in seconds: 61.375491166953
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Load(C3, P3, ATL)
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Fly(P3, ATL, SFO)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)
Unload(C3, P3, SFO)



### 2) - astar_search h_ignore_preconditions


In [7]:
%run run_search.py -p 2 -s 9


Solving Air Cargo Problem 2 using astar_search with h_ignore_preconditions...

Expansions   Goal Tests   New Nodes
   1506        1508       13820   

Plan length: 9  Time elapsed in seconds: 20.807023262954317
Load(C1, P1, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Load(C3, P3, ATL)
Fly(P3, ATL, SFO)
Unload(C3, P3, SFO)



### 3) - astar_search h_pg_levelsum


In [8]:
%run run_search.py -p 2 -s 10


Solving Air Cargo Problem 2 using astar_search with h_pg_levelsum...

Expansions   Goal Tests   New Nodes
    86          88         841    

Plan length: 9  Time elapsed in seconds: 321.5026430550497
Load(C1, P1, SFO)
Fly(P1, SFO, JFK)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Load(C3, P3, ATL)
Fly(P3, ATL, SFO)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)
Unload(C3, P3, SFO)



### Third problem: air_cargo_p3

- Search methods: astar_search h_1, astar_search h_ignore_preconditions and astar_search h_pg_levelsum

### 1) - astar_search h_1

In [9]:
%run run_search.py -p 3 -s 8


Solving Air Cargo Problem 3 using astar_search with h_1...

Expansions   Goal Tests   New Nodes
  18151       18153       159038  

Plan length: 12  Time elapsed in seconds: 504.1941834639292
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P1, SFO, ATL)
Load(C3, P1, ATL)
Fly(P2, JFK, ORD)
Load(C4, P2, ORD)
Fly(P2, ORD, SFO)
Fly(P1, ATL, JFK)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)
Unload(C3, P1, JFK)
Unload(C4, P2, SFO)



### 2) - astar_search h_ignore_preconditions


In [10]:
%run run_search.py -p 3 -s 9


Solving Air Cargo Problem 3 using astar_search with h_ignore_preconditions...

Expansions   Goal Tests   New Nodes
   5118        5120       45650   

Plan length: 12  Time elapsed in seconds: 130.61921787296887
Load(C1, P1, SFO)
Fly(P1, SFO, ATL)
Load(C3, P1, ATL)
Fly(P1, ATL, JFK)
Unload(C1, P1, JFK)
Load(C2, P2, JFK)
Fly(P2, JFK, ORD)
Load(C4, P2, ORD)
Fly(P2, ORD, SFO)
Unload(C2, P2, SFO)
Unload(C3, P1, JFK)
Unload(C4, P2, SFO)



### 3) - astar_search h_pg_levelsum


### Summary from results


#### Air Cargo Problem 1

|   |Expansions   |  Goal Tests |  New Nodes  | Plan length |Time elapsed in seconds |
|---|---          |---          |---          |---          |---                    |
|astar_search h_1   | 55  | 57  |224   | 6 | 0.0789819739293307
| astar_search h_ignore_preconditions  |  41 |  43 | 170  | 6 | 0.08492878801189363
|astar_search h_pg_levelsum   | 11  | 13  | 50  | 6 | 3.08339868998155


#### Air Cargo Problem 2

|   |Expansions   |  Goal Tests |  New Nodes  | Plan length |Time elapsed in seconds |
|---|---          |---          |---          |---          |---                    |
|astar_search h_1   | 4853  | 4855  | 44041   | 9 | 61.375491166953
| astar_search h_ignore_preconditions  |  1506 |  1508 | 13820  | 9 | 20.807023262954317
|astar_search h_pg_levelsum   | 86  | 88  | 841  | 9 | 321.5026430550497

                                                     
#### Air Cargo Problem 3

|   |Expansions   |  Goal Tests |  New Nodes  | Plan length |Time elapsed in seconds |
|---|---          |---          |---          |---          |---                    |
|astar_search h_1   | 18151  | 18153  | 159038   | 12 | 504.1941834639292
| astar_search h_ignore_preconditions  |  5118 |  5120 | 45650  | 12 | 130.61921787296887
|astar_search h_pg_levelsum   | 404  | 406  | 3718  | 12 | 2100.5661585619673